### ISSUE TITLE
Support connecting to Tor hidden services

### ISSUE THREAD

**Issue by FedericoCeratto** [bitfireAT/davdroid#407]
Related to #231
Supporting connections to Tor hidden services could make the deployment of personal CalDAV servers easier:
- The hidden service name can act as a certificate
- The server could be run behind NAT and/or moved to different network locations easily
 
Thank you!

**Comment by rfc2822** [comment]
Does the SOCKS proxy approach of #231 work?

**Comment by FedericoCeratto** [comment]
I'm don't know how to set those variables to testing. In order to work, davdrod will have to perform DNS resolution through the proxy as well.

**Comment by rfc2822** [comment]
Can you provide me with a tutorial on how to use Tor with Android and how other apps handle this? I don't use Tor at all.

**Comment by FedericoCeratto** [comment]
This is a bit lengthy https://trac.torproject.org/projects/tor/wiki/doc/TorifyHOWTO - but in essence the trick is to support a proxy, typically on port 8188: https://guardianproject.info/apps/orbot/
Ideally the user should be able to enable/disable the proxy from the application.

**Comment by rnauber** [comment]
Hi, 
I hacked together some quick and dirty tor support. Please test it and give me some feedback...
#421

**Comment by FedericoCeratto** [comment]
Sorry @rnauber the second link is broken.

**Comment by rnauber** [comment]
Hi there,
I updated my pull request so it does not break non-tor addresses any more: #449 
So please test!
(A prebuild binary can be found here:
https://github.com/rnauber/davdroid/releases/download/torv2/davdroid_torv2_debug.apk
)

**Comment by FedericoCeratto** [comment]
@rnauber - it works, thank you very much for the patch!

**Comment by rnauber** [comment]
Hi,
i updated the binaries once more:
https://github.com/rnauber/davdroid/releases/tag/v0.7.1_tor

I would love to have this patch going upstream, what is necessary to make that happen? 

See you,
Richard

**Comment by rfc2822** [comment]
I don't see why this is necessary. I have tried this successfully with Android 4.4:

1. Install [Orbot](https://www.torproject.org/docs/android.html.en)
1. For WiFi, set HTTP proxy localhost:8118.
1. For mobile data, set HTTP proxy localhost:8118 (Mobile networks / Access Points Names (APN) / <your access point> / proxy: localhost, port: 8118).
1. All apps which take the system HTTP proxy settings into account, including DAVdroid, connect via Orbot und thus TOR.

Because all HTTP requests are sent via Orbot, domain resolving is not an issue and .onion etc. work. The only thing that won't work are SRV records with .onion domains.

What do you think? What exactly is the patch needed for?

**Comment by rfc2822** [comment]
Please also have a look at https://davdroid.bitfire.at/faq/entry/using-with-tor

**Comment by rnauber** [comment]
Hi,
the use-cases I would like to have supported are (in order of importance):
1. using .onion urls in davdroid without affecting the whole system ("why is my internet so slow?") or rooting the phone 
2. mixing tor and non-tor servers in davdroid, while using their respective connections

The patch meets both requirements and works without further configuration for the default orbot settings. But an per-app configuration of the proxy settings in davdroid would fine with me too...

Btw., thanks for creating this great tool!
Richard

**Comment by rfc2822** [comment]
Ok, this sounds like a very specific thing. Basically, DAVdroid is supporting Tor. Do you use Tor only to overcome NAT issues? Maybe IPv6 or VPN would be a better solution.

**Comment by FedericoCeratto** [comment]
Same as @rnauber:
- many users don't want to root their phones. In some countries it is even illegal
- enabling proxying globally will impact every network use, including web browsing
- also, the user has to configure the proxy manually on every new wifi network

**Comment by rfc2822** [comment]
> many users don't want to root their phones. In some countries it is even illegal

Why root the phone?

> enabling proxying globally will impact every network use, including web browsing

Isn't that the reason why you want to use Tor, to hide your traces, regardless of which app you use?

> also, the user has to configure the proxy manually on every new wifi network

Makes sense.

**Comment by rnauber** [comment]
Let me try to answer that:

>Why root the phone?

With a rooted phone you can just use the orbot transparent proxy feature to route selected apps (e.g. davdroid) through tor.

>>enabling proxying globally will impact every network use, including web browsing

>Isn't that the reason why you want to use Tor, to hide your traces, regardless of which app you use?

Yes, but there are gradations in privacy demands of different users e.g.:

* using tor just as DNS replacement where ipv6 is unavailable and you want to have an easy self-hosted server for davdroid. All other apps use clearnet access.

* using tor for davdroid and all apps, except youtube, because it lags otherwise

* using tor for everything, changing the circuit every minute

>>also, the user has to configure the proxy manually on every new wifi network

>Makes sense.

Yes, this one is a big hurdle for a hassle free setup...

**Comment by rfc2822** [comment]
If you really want to use Tor because you transmit "dangerous" content, you probably won't use Youtube and don't connect to many WiFi networks.

But I see that this is an enhancement request for very special cases, while DAVdroid basically supports Tor and hidden services.

**Comment by rnauber** [comment]
> If you really want to use Tor because you transmit "dangerous" content, you probably won't use Youtube and don't connect to many WiFi networks.

I agree with the former but not necessarily with the latter and think Tor should be used for non-"dangerous"  content as well. And to broaden its user base, tor-hosted servers/apps have to be really simple to set up (ideally, even simpler than the non-privacy-conserving alternatives, e.g. dyndns).

So the consensus is to have a http-proxy option in the DavDroid settings, right? 
I suggest to have the standard orbot values (localhost:8118) set as default.

So, how to proceed on this? If none else volunteers, I could try to write a patch that adds the UI. But it will probably take me a while...

See you,
Richard

**Comment by rnauber** [comment]
Hi,
I started to click together a proxy-settings UI for Davdroid supporting HTTP-proxies. But  the HTTP proxy  of Orbot does not support the HTTP requests we use (http://www.pps.univ-paris-diderot.fr/~jch/software/polipo/manual/POST-and-PUT.html), therefore **only HTTPS DAV servers will work, no HTTP ones!**
So it would be nice, if you could state that in the FAQ:
https://davdroid.bitfire.at/faq/entry/using-with-tor

But I think the usecase 
DavDroid --> TOR --> **http**://davserver_xyz.onion
is very much worth supporting, because of the extremely easy set up of a self-hosted server (there is no need to generate certificates, end-to-end encryption is seamlessly provided by tor...).

So I think we have three options here:
1) Add a SOCKS proxy configuration option to DavDroid
2) Add a HTTP proxy configuration to DavDroid that supports the CONNECT tunnelling
(But this is a very indirect approach, it would be much nicer to go for 1) instead)
3) drop this usecase entirely

I am very much in favour of 1) but what do you think (especially @rfc2822 )?

See you,
Richard
